### DISCUSSION TITLE
Refer to instance of class in metaclass

### DISCUSSION THREAD

**Discussion by henribru** [python/typing#1507] [Q&A] [answered]
In a meta class, is there a way to refer to an instance of the class that uses the metaclass? In a regular classmethod you can use `Self`, but in a meta class `Self` refers to the class itself, so that doesn't work:

https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmB9ApgDxllAZyTCKiQgTBBigGUsAbYAWACh2BjRgQ0MKgBZLPx5oswmDwAU8BFgCUALnZQ1UACZZgUBACNp3QgBooYPQCsl9JsAVQAtAD4bzawDpP7Lr35CRhGJY0hBYUtx8hAC8wqLikjzKUJ7u3hxsIFgAblg8jBhywbGB4u76CuyZOXkFiEUBQWUGxUHSCu1AA

https://mypy-play.net/?mypy=latest&python=3.11&gist=ac7550a5b678afea19f55d2d92e6eec8 (Mypy actually errors on the usage of `Self` in the meta class as well, though the result still matches Pyright)

**Comment by Gobot1234** [comment, marked as answer]
This is something that I considered when writing PEP 673 but the idea never went anywhere. However in your case you can probably just use a type var bound to Message and take that and return it from .pb

**Reply by henribru** [reply]
Ah, you're right of course, `def pb(cls: type[M], obj: M) -> M: ...` works like a charm. Thank you!
